Enhance Ruby Testing Guide with setup and execution details

aguspe opened the pull request:
### **User description**
Expanded the Ruby Testing Guide to include local development setup, test running methods, and environment variables.

<!-- Thanks for Contributing to Selenium! -->
<!-- Please read our contribution guidelines: https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md -->

### 🔗 Related Issues
<!-- Example: Fixes #1234 or Closes #5678 -->
<!-- If the reason for this PR is not obvious, consider creating an issue for it first -->

### 💥 What does this PR do?
<!-- Describe what this change includes and how it works -->

### 🔧 Implementation Notes
<!--- Why did you implement it this way? -->
<!--- What alternatives to this approach did you consider? -->

### 💡 Additional Considerations
<!--- Are there any decisions that need to be made before accepting this PR? -->
<!--- Is there any follow-on work that needs to be done? (e.g., docs, tests, etc.) -->

### 🔄 Types of changes
<!-- ✂️ Please delete anything that doesn't apply -->
- Cleanup (formatting, renaming)
- Bug fix (backwards compatible)
- New feature (non-breaking change which adds functionality *and tests!*)
- Breaking change (fix or feature that would cause existing functionality to change)


___

### **PR Type**
Documentation


___

### **Description**
- Expanded Ruby Testing Guide with local development setup instructions

- Added comprehensive RSpec and Rake test execution methods

- Introduced Code Style & Linting and Documentation generation sections

- Enhanced Guards section with conditions table and debugging tips

- Reorganized content for better clarity and workflow guidance


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["Ruby Testing Guide"] --> B["Local Development Setup"]
  A --> C["Test Execution Methods"]
  C --> C1["Bazel CI Workflow"]
  C --> C2["RSpec Local Workflow"]
  C --> C3["Rake Shortcuts"]
  A --> D["Code Style & Linting"]
  A --> E["Documentation Generation"]
  A --> F["Enhanced Guards & Debugging"]
  A --> G["Environment Variables"]
```



<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>TESTING.md</strong><dd><code>Comprehensive expansion of Ruby testing documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/TESTING.md

<ul><li>Added "Local Development Setup" section with <code>bundle install</code> <br>instructions<br> <li> Expanded "Running Tests" section with three distinct approaches: <br>Bazel, RSpec, and Rake<br> <li> Added new "Code Style & Linting" section documenting Rubocop usage<br> <li> Added new "Documentation" section for YARD documentation generation<br> <li> Enhanced "Guards" section with detailed conditions table and guard <br>examples<br> <li> Added "Debugging" subsection with breakpoint instructions<br> <li> Reorganized "Helpers" section with improved table formatting<br> <li> Restructured "Environment Variables" section with clearer examples<br> <li> Improved overall formatting with better markdown tables and code <br>blocks</ul>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16805/files#diff-bad8c461d37ea389c61053ab7010cf5d1103371f46c00422b2b63dd41d89ce47">+133/-80</a></td>

</tr>
</table></td></tr></tbody></table>

</details>

___

aguspe committed:
Enhance Ruby Testing Guide with setup and execution details

Expanded the Ruby Testing Guide to include local development setup, test running methods, and environment variables.

titusfortner commented on rb/TESTING.md:
```diff
@@ -1,13 +1,24 @@
+
+## Local Development Setup
+
+Before running tests, navigate to the `rb/` directory and install the required dependencies.
```
local development still requires running bazel to generate the atoms and devtools code

titusfortner replied:
I'm not sure this is actually up to date: https://github.com/SeleniumHQ/selenium/blob/trunk/README.md#ruby

But that info should be moved into this file.

titusfortner commented on rb/TESTING.md:
```diff
@@ -24,34 +35,81 @@ module Selenium
+rake spec:unit          # Run unit tests
+rake spec:integration   # Run integration tests (Chrome default)
+rake docs               # Generate docs
+rake clean              # Clean generated artifacts
```
Haven't looked at the code in a while. Does clean require we re-run the bazel commands.

aguspe replied:
Rake clean doesn't apply here, and I will actually remove it because the clean task is on the rakefile at the selenium level and not under rb/rakefile

titusfortner commented on rb/TESTING.md:
```diff
@@ -24,34 +35,81 @@ module Selenium
+| `exclusive` | Test is skipped entirely if conditions not met (not pending). |
+| `exclude` | Test is skipped (use for broken/unreliable tests). |
+
+### Guard Conditions
```
Link to the spec_helper where these are defined

titusfortner commented on rb/TESTING.md:
```diff
@@ -69,56 +127,57 @@ end
-| `headless` | `true`, `false` |
-| `bidi` | `true`, `false` |
-| `ci` | `true`, `false` |
+Selenium enforces strict code style using **Rubocop**. CI will fail if linting errors are present.
```
Point to the rubocop.yml and I think we should say to prefer updating that to in-file guards. Not sure your preference.

titusfortner reviewed (changes requested):
This is great. A few comments and requests for clarification. Thanks!

aguspe committed:
Add Steep type signature documentation for Ruby contributors

- Added comprehensive Type Signatures with Steep section
- Includes instructions for adding RBS files for new classes
- Added guide for updating existing type signatures
- Documented Steep commands for type checking
- Added best practices and common RBS types reference table
- Added RubyMine IDE setup instructions
- Enhanced debugging section with Bazel and RSpec options
- Added interactive REPL usage with bazel run //rb:console

aguspe committed:
comments addressed

aguspe committed:
Merge branch 'testing_guides' into aguspe-patch-1

luke-hill commented on rb/TESTING.md:
````diff
@@ -1,13 +1,44 @@
+
+```shell
+cd rb
+bundle install
````
Do you wanna add the bundler install, maybe with a desireable v4 one?

luke-hill commented on rb/TESTING.md:
```diff
@@ -31,27 +62,54 @@ end
-| `only` | Test is pending if conditions are NOT met |
-| `exclusive` | Test is skipped entirely if conditions not met |
-| `exclude` | Test is skipped (use for broken/unreliable tests) |
+| --- | --- |
```
RM should auto format all this stuff for you to avoid you getting into this mess where things flag

luke-hill reviewed (commented):
Really good - lots of this is well documented and much easier to comprehend
